src/tcache.c

```diff
@@ -576,19 +585,34 @@ tcache_init(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
 		if (i < SC_NBINS) {
 			tcache_slow->lg_fill_div[i] = 1;
 			tcache_slow->bin_refilled[i] = false;
+			tcache_slow->bin_flush_delay_items[i]
+			    = tcache_gc_item_delay_compute(i);
 		}
 		cache_bin_t *cache_bin = &tcache->bins[i];
 		cache_bin_init(cache_bin, &tcache_bin_info[i], mem,
 		    &cur_offset);
 	}
+	/*
+	 * For small size classes beyond tcache_maxclass (i.e. nhbins < NBINS),
+	 * their cache bins are initialized to a state to safely and efficiently
+	 * fail all fastpath alloc / free, so that no additional check around
+	 * nhbins is needed on fastpath.
+	 */
+	for (unsigned i = nhbins; i < SC_NBINS; i++) {
+		/* Disabled small bins. */
```

Comment:
Yeah agreed it's best to keep it NULL / 0 all the way. The only reason that I have it initialized awkwardly this way is, we access it on cache_bin_alloc_easy here: https://github.com/jemalloc/jemalloc/blob/6cdac3c573de86c8d59d69fca8f1778bdbec25e0/include/jemalloc/internal/cache_bin.h#L247

and I don't really want to check `nhbins` on the fast path, which means all fast path will go down the cache_bin_alloc path and peek into the stack (although the result won't be used, like the empty case). Previously we have seen "touching" an invalid pointer that way caused unexpected slowdowns. I couldn't find better tricks to avoid the branch on fast path so here we are...

If I'm calculating correctly, the stack_head of the disabled bins should all point to 0x7a7a... junk we padded there, so hopefully still crash soon enough to avoid confusion.

Reply:
Makes sense, disregard.